[webnn] Add tests for concat operation.

BruceDai opened the pull request:
@fdwr This pr only has `concat` operation tests, this would make you review easily. I moved tests and test data onto `resources/test_data/concat.json` JSON file which will make maintain test smoothly,  and I also addressed some feedbacks which you commented on #34287. Please help review it, thanks a lot. 

@dontcallmedom @Honry @huningxin Please also take a review, thanks.

BruceDai committed:
[webnn] Add tests for concat operation.

BruceDai commented on webnn/resources/test_data/concat.json:
```diff
@@ -0,0 +1,632 @@
+    }
+  },
+  "inputsData": {
+    "float64Data": [
```
These `inputsData` and below `expectedData` were automatically generated by [this scripts](https://github.com/webmachinelearning/webnn-baseline/pull/22) of WebNN-Baseline project.

BruceDai committed:
[webnn] Updated concat tests to align with latest WebNN API Spec
        createContext() / createContextSync()
        build() / buildSync()
	compute() / computeSync()

fdwr commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,72 @@
+    graph = await builder.build({outputOperand});
+    await context.compute(graph, namedInputs, outputs);
+  }
+  assert_array_approx_equals_ulp(outputs.outputOperand, expectedShapeValue.data, PrecisionMetrics.concat.ULP[operandType], operandType);
```
[minor] Hmm, why does this function name use snake case whereas nearly every other function and variable uses camelCase? (e.g. sizeOfShape, computeSync, loadTestData)?

BruceDai replied:
The assertion function `assert_array_approx_equals_ulp()` was intended to add into [`testharness.js`](https://github.com/web-platform-tests/wpt/blob/master/resources/testharness.js) previously, so its' name followed like [`function assert_array_equals()`](https://github.com/web-platform-tests/wpt/blob/master/resources/testharness.js#L1614) of  [`testharness.js`](https://github.com/web-platform-tests/wpt/blob/master/resources/testharness.js). 
Since `assert_array_approx_equals_ulp()` is especial for WebNN API tests, I'm OK to use camelCase style.

fdwr replied:
I see. So there was some external convention.

BruceDai replied:
Yes. So what's your suggestion for this assertion function? Thanks.

fdwr replied:
Oh, I generally lean toward internal consistency with whatever code is immediately around it, but there's also the consistency argument of matching other "assert_" functions. I'm fine with whichever you decide - you answered my question either way. ✅

BruceDai replied:
I leave this assert to be consistent with other "assert_" functions in testharness.js, thanks.

fdwr commented on webnn/resources/utils.js:
```diff
@@ -34,28 +43,52 @@ const DeviceTypeArray = ['cpu', 'gpu'];
+    distance = actualBitwise - expectedBitwise;
+    distance = distance >= 0 ? distance : -distance;
+    assert_true(distance <= nulp,
+                `actual ${actual[i]} should be close enough to expected ${expected[i]} by the acceptable ${nulp} ULP distance, while they have ${distance} ULP distance`);
```
Propose "while" -> "but".

BruceDai replied:
Updated, thanks.

fdwr commented on webnn/resources/utils.js:
```diff
@@ -5,24 +5,33 @@ const ExecutionArray = ['async', 'sync'];
 
+// Refer to precision metrics on https://github.com/webmachinelearning/webnn/issues/265#issuecomment-1256242643
+const PrecisionMetrics = {
+  concat: {ULP: {float32: 0, float16: 0}},
```
Nice.

fdwr commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,72 @@
+        }
+        const expectedShapeValue = {shape: test.expected.shape, data: expectedData[expectedDataCategory]};
+        await testConcat(operandType, isSync, inputShapeValues, test.axis, expectedShapeValue);
+      }, `${test.name} / ${operandType} / ${deviceType} / ${executionType}`);
```
[ignorable] I notice you usually put a new line after `}`'s, and so at first glance, it looked like a typo here, as I couldn't tell if the brace was in the wrong place for the `for` loop, not realizing it was a second parameter to the promise call. So new line after the `},`? (note I'm not familiar with Javascript common style conventions, but some I find easier to visually parse than others 🤷)

fdwr commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,72 @@
+        for (const shape of inputShapes) {
+          const size = sizeOfShape(shape);
+          inputShapeValues.push({shape, data: inputsData[inputDataCategory].slice(pos, pos + size)});
+          pos += size;
```
How about `position` or `offset`? (I just like full words :))

BruceDai replied:
Updated, thanks.

fdwr commented on webnn/resources/utils.js:
```diff
@@ -34,28 +43,52 @@ const DeviceTypeArray = ['cpu', 'gpu'];
   }
+
+  const json = loadJSON(file);
+  return JSON.parse(json.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m));
```
Nice, this way we can have comments in the JSON.

Honry commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,72 @@
+  } else {
+    graph = await builder.build({outputOperand});
+    await context.compute(graph, namedInputs, outputs);
+  }
```
Is that possible to wrap this code snippet into a common function? Which could be shared to other tests.
````
  let graph;
  if (syncFlag) {
    graph = builder.buildSync({outputOperand});
    context.computeSync(graph, namedInputs, outputs);
  } else {
    graph = await builder.build({outputOperand});
    await context.compute(graph, namedInputs, outputs);
  }
```

BruceDai replied:
I extracted them into `buildAndCompute()` function of `utils.js`, and also did for the codes of creating context and builder by `createContextAndBuilder()` function, thanks.

Honry reviewed (approved):
Generally LGTM, thanks @BruceDai !

BruceDai committed:
[webnn] Optimize concat tests by addressing comments.

BruceDai reviewed (commented):
@fdwr @honry Thanks for your reviewing, I've updated PR to address your comments, please take another look, thanks.

BruceDai committed:
[webnn] Align the usage for sync and async functions.

BruceDai commented:
Hi @huningxin, according to your suggestion on https://github.com/webmachinelearning/webnn-polyfill/pull/172, I updated this PR by invoking sync functions and async functions clearly not using `syncFlag` flag, please have a review, thanks.

BruceDai committed:
[webnn] Optimize concat tests with renaming variables.
         inputDataCategory -> inputDataSource
	 expectedDataCategory -> expectedDataSource

BruceDai committed:
[webnn] Optimized concat tests by moving if condition checking out of for loop.

BruceDai committed:
[webnn] Refined code to reuse for other operation tests.

BruceDai commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,48 @@
+  return [{outputOperand}, inputs, outputs];
+};
+
+testWebNNOperation('concat', concatTests(), buildGraph);
```
Hi @fdwr, @Honry and @huningxin,
I refined code to reuse for later adding other operations tests. Now each test file named \<operation-name\>.https.any.js will have `<operation-name>Tests()`, `buildGraph()` two functions, and run test by invoking `testWebNNOperation()` function, please take another look, thanks.

BruceDai commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,307 @@
+            }
+            tolerance++;
+          }
+          break;
```
@fdwr I created this `getPrecisonTolerance()` function, and add detail implementation to get a ULP tolerance for these operations which need compute by IEPOE, please take another look, thanks.

fdwr replied:
Would it be simpler to just include the function name in the table, right beside all the other values, rather than needing to fill in a switch statement (because then there are two separate places to look for values and update, a table and a code switch)? e.g.

```
function getGemmPrecisionTolerance(shapeA, options) {
  const shapeA = resources.inputA.shape;
  options = resources.options;
  ...
}

...

const PrecisionMetrics = {
    batchNormalization: {ULP: {float32: 6, float16: 6}},
...
    gemm: {ULP: {float32: getGemmPrecisionTolerance, float16: getGemmPrecisionTolerance}}
...
    maxPool2d: {ULP: {float32: 0, float16: 0}},
...
};

...

function getPrecisonTolerance(opName, metricType, tensorDataType, resources) {
   let tolerance;
    if (metricType === 'ULP') {
      tolerance = PrecisionMetrics[opName].ULP[tensorDataType];
    }
    else if (metricType === 'ATOL') {
      tolerance = PrecisionMetrics[opName].ATOL[operandType];
    }

    // If the tolerance is dynamic, then evaluate the function to get the value.
    if (x instanceof Function) {
      tolerance = tolerance(resources);
    }
    return tolerance;
}
```

(but I'm also fine with the current code)

BruceDai replied:
Updated, thanks.

Honry commented on webnn/resources/utils.js:
```diff
@@ -93,52 +35,288 @@ function loadTestData(file) {
+   * @param {Boolean} [options.bTranspose]
+   * @returns {Number} A tolerance number
+   */
+  function getGemmPrecisionTolerance(shapeA, options) {
```
Is there any reason that we have to put these functions inside the `testWebNNOperation`? This would make code not easy to read.

BruceDai replied:
Thanks for your suggestion, moved these functions outside.

fdwr commented on webnn/resources/utils.js:
```diff
@@ -93,52 +35,288 @@ function loadTestData(file) {
+  const checkResults = (opName, operandType, namedOutputOperands, outputs, resources) => {
+    const metricType = PrecisionMetrics[opName] ? Object.keys(PrecisionMetrics[opName])[0] : 'ULP';
+    const description = `test ${opName} ${operandType}`;
+    const tolerance = getPrecisonTolerance(opName, metricType, operandType, resources);
```
It's possible someday that WebNN will support operators which return *multiple* outputs, and each output could have its own tolerance. It's not an issue *yet* for WebNN v1, and so I'm not worried about it, but we should keep that in mind when writing the code. The only operators I can think of right *now* are MAX_POOL and TOP_K, but those both expect exact results, and so both outputs have the same tolerance anyway.

BruceDai committed:
[webnn] Address comments to move nested functions outside.

BruceDai commented on webnn/resources/utils.js:
```diff
@@ -5,76 +5,18 @@ const ExecutionArray = ['sync', 'async'];
-
 const TypedArrayDict = {
   float32: Float32Array,
 };
```
Added these code, thanks.

BruceDai reviewed (commented):
Thanks @Honry @fdwr, I've updated the commit to address your comments, please take another look.

fdwr commented:
> Thanks @Honry @fdwr, I've updated the commit to address your comments, please take another look.

Push the new changes?

![image](https://user-images.githubusercontent.com/1809166/202694027-edd48256-9371-4b26-ac0f-79218af1dd08.png)

BruceDai commented:
@fdwr Sorry. Updated, thanks.

fdwr reviewed (approved):
Looks nice. Thanks.

huningxin commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,48 @@
+  return [{outputOperand}, inputs, outputs];
+};
+
+testWebNNOperation('concat', concatTests(), buildGraph);
```
Could this be something like
```js
testWebNNOperation('concat', '/webnn/resources/test_data/concat.json', (inputOperands, resources) => {
    return {outputOperand: builder.concat(inputOperands, resources.axis)}; });
```

huningxin reviewed (commented):
Looks good with some nits, thanks!

BruceDai commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,48 @@
+
+// https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-concat
+
+const concatTests = () => {
```
Hi @huningxin, thanks for your reviewing and suggestions.
Since there're different implementations (L#16 to L#29) to set `targetTests` for WebNN operations, for example the implementation for `gemmTests` likes:
```js
  for (const test of tests) {
    const gemmOptions = {};
    const options = test.options;
    if (options !== undefined) {
      if (options.c !== undefined) {
        gemmOptions.c = {data: inputsData[options.c.data]};
        if (options.c.shape !== undefined) {
          gemmOptions.c['shape'] = options.c.shape;
        }
      }
      if (options.alpha !== undefined) {
        gemmOptions.alpha = inputsData[options.alpha.data];
      }
      if (options.beta !== undefined) {
        gemmOptions.beta = inputsData[options.beta.data];
      }
      if (options.aTranspose !== undefined) {
        gemmOptions.aTranspose = options.aTranspose;
      }
      if (options.bTranspose !== undefined) {
        gemmOptions.bTranspose = options.bTranspose;
      }
    }
    targetTests.push(
      {
        name: test.name,
        operandType: test.type,
        inputA: {shape: test.inputs.a.shape, data: inputsData[test.inputs.a.data]},
        inputB: {shape: test.inputs.b.shape, data: inputsData[test.inputs.b.data]},
        expected: {shape: test.expected.shape, data: {outputOperand: expectedData[test.expected.data]}},
        options: gemmOptions
      }
    );
  }
```
and implementation for `clampTests` likes:
```js
  for (const test of tests) {
    const expectedDataSource = test.expected.data;
    const inputDataSource = test.input.data;
    targetTests.push(
      {
        name: test.name,
        operandType: test.type,
        input: {shape: test.input.shape, data: inputsData[inputDataSource]},
        expected: {shape: test.expected.shape, data: {outputOperand: expectedData[expectedDataSource]}},
        options: test.options
      }
    );
  }
```

 so I implemented each `<operation-name>Tests` function for each operation, thanks.

BruceDai replied:
@huningxin I just submitted tests for clamp (#37067) / gemm (#37068) / matmul (#37069) / reshape (#37070) / relu (#37071) / slice (#37072) operations, PTAL, thanks.

huningxin replied:
This function looks like mainly setting up the test object for `targetTests`. It seems to me that all information and configuration of each test object come from the JOSN file. I am wondering: could we just define all these info in the JSON file and load it into `resources`. I suppose we don't need different JS implementation to setup test object for different ops.

Did I miss anything?

BruceDai replied:
@huningxin Do you suggest add actual test data for "data"? For example https://github.com/web-platform-tests/wpt/blob/32290b240b50ef914a03af71f9136b805d177b10/webnn/resources/test_data/concat.json#L8

my purpose of using the index string key of input / expected data is to reuse test data, thanks.

BruceDai replied:
@huningxin I've updated this PR for more reuse code, please take another look. And there's a [reusable tests demo](https://github.com/BruceDai/wpt/pull/442/files) on my own repo for your quickly review, thanks.

BruceDai commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,48 @@
+    let position = 0;
+    for (const shape of inputShapes) {
+      const size = sizeOfShape(shape);
+      inputShapeValues.push({shape, data: inputsData[inputDataSource].slice(position, position + size)});
```
This code is especial to `concat` op. To inputs of others operations, it's simply to set `data`. @huningxin

BruceDai commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,48 @@
+  for (let i = 0; i < inputShapeValues.length; i++) {
+    inputOperands.push(builder.input('input' + i, {type: operandType, dimensions: inputShapeValues[i].shape}));
+    inputs['input' + i] = new TestTypedArray(inputShapeValues[i].data);
+  }
```
@huningxin Thanks for your suggestions, I will try to optimize setting `inputOperands`, `inputs` and `outputs` in the next PR.

BruceDai reviewed (commented):
@huningxin Thanks for your reviewing and suggestions, I will follow up in next PR, thanks.

BruceDai committed:
[webnn] More refined code to reuse for other operation tests.

huningxin commented on webnn/resources/test_data/concat.json:
```diff
@@ -2,662 +2,1712 @@
+          0.1211843192577362
+        ]
       },
       "type": "float32"
```
You may need to set type for each operand. Because some operations may have different types for different input and output operands.

BruceDai replied:
Thanks, I will update it. And is it ok to still keep "type" here to indicate that current test is for float32 precision?

huningxin replied:
What's the use case of it? The spec only defines operand data type, not a operation data type.

BruceDai replied:
Regarding to your last comment "some operations may have different types for different input and output operands.", which is the main precision target. 
This helper "type" would clearly show the test precision target.

huningxin replied:
I understand the precision target would be represented by the data type of the expected output data. That is accurate and will be used in the test.

Unless this helper "type" has a specific use case, I'd recommend to remove it.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -34,34 +34,59 @@ const loadTestData = (file) => {
+    aTranspose: false,
+    bTranspose: false
+  }
+  const options = resources.options ? resources.options : defaultOptions;
```
The default options are defined by spec and should be implemented by browser. 
spec: https://www.w3.org/TR/webnn/#dictdef-mlgemmoptions
chromium implementation: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.idl;l=27

huningxin replied:
What's the use case of getting the default settings? If user code doesn't set the value of an option, the default value will be used if it is defined by the spec.

huningxin replied:
If you have concerns, I'd recommend to leave this function to following PRs. It would be good to have only code needed for this concat test case.

BruceDai replied:
OK, I will move it, and add it in next PR, thanks.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -208,86 +230,159 @@ const assert_array_approx_equals_ulp = (actual, expected, nulp, dataType, descri
+ * @param {String[]} nameArray
+ * @returns {MLOperand[]} Input operands
+ */
+const createInputOperands = (builder, resources, nameArray) => {
```
Should this method be used by concat test? It is not used by anyone.

BruceDai replied:
Since the inputs of `concat` is a sequence, it's special to other ops.
I form its inputs, like

```
      "inputs": [
        {
          "name": "input1",
          "shape": [12],
          "data": [
            -0.39444134019222243,
            ...
          ]
        },
        {
          "name": "input2",
          "shape": [12],
          "data": [
            0.4918989118791477,
            ....
          ]
        }
      ],
```
while for inputs of `clamp`, I form it like
```
      "inputs": {
        "x": { // use "x" for input operand name
            "shape": [12],
            "data": [
              -0.39444134019222243,
              ...
          -  ]
          }
        }
```
so this `createInputOperands` is for other ops, please see the usage on [demo tests](https://github.com/BruceDai/wpt/pull/442/files) .

huningxin replied:
If this function is not used in this PR, I'd suggest to leave it to following PRs where it is used.

huningxin commented on webnn/concat.https.any.js:
```diff
@@ -7,42 +7,13 @@
-  const inputShapeValues = resources.inputShapeValues;
-  const operandType = resources.operandType;
-  const TestTypedArray = TypedArrayDict[operandType];
+const computeConcat = (builder, resources) => {
```
This is not a compute function. I recommend to rename it to `buildConcat`.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -208,86 +230,159 @@ const assert_array_approx_equals_ulp = (actual, expected, nulp, dataType, descri
+ * Build a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} computeFunc - A compute function
```
The name is confusing. It is actual a build function that build a specific operation.

BruceDai replied:
Thanks, I will update it.

huningxin reviewed (commented):
@BruceDai , thanks for the update. It's looking better.

BruceDai reviewed (commented):
@huningxin Thanks for your reviewing and suggestions.

BruceDai commented:
@huningxin I've updated PR to address your comments, please take another look, thanks.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,312 @@
+
+/**
+ * Run WebNN operation tests.
+ * @param {String} name - An operation name
```
Should this parameter be named `operationName`? `name` is not clear about whose name it is. Please make this one and all the other `name`s to be more specific.

BruceDai replied:
Updated, thanks.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,312 @@
+ * @returns [namedOperands, inputs, outputs]
+ */
+const buildGraph = (builder, resources, buildFunc, name) => {
+  const outputOperand = buildFunc(builder, resources, name);
```
I don't find the parameter `name` is used by this method `buildGraph()` and `buildFunc()`. It probably could be removed.

BruceDai replied:
Removed it firstly, add this parameter later in next PR, thanks.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,312 @@
+  } else {
+    const outputName = resources.expected.name ? resources.expected.name : 'output';
+    // matmul 1D with 1D produces a scalar which doesn't have its shape
+    const shape = resources.expected.shape ? resources.expected.shape : [];
```
`[]` is not a valid dimensions, it'd be better to use `[1]` instead.

BruceDai replied:
Updated, thanks.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,312 @@
+    for (let i = 0; i < resources.expected.length; i++) {
+      const subExpected = resources.expected[i];
+      outputs[subExpected.name] = new TypedArrayDict[subExpected.type](sizeOfShape(subExpected.shape));
+      namedOperands[subExpected.name] = outputOperand[i];
```
It'd be better to let `buildFunc()` return `namedOperands`. The mapping of `outputOperand[i]` to `namedOperands[subExpected.name]` is unclear.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,312 @@
+    // matmul 1D with 1D produces a scalar which doesn't have its shape
+    const shape = resources.expected.shape ? resources.expected.shape : [];
+    outputs[outputName] = new TypedArrayDict[resources.expected.type](sizeOfShape(shape));
+    namedOperands[outputName] = outputOperand;
```
ditto. Consider letting `buildFunc()` return `namedOperands`.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,312 @@
+ * @param {String} name - An output name
+ * @returns {Number[]} An expected data array
+ */
+const getExpectedData = (resources, name) => {
```
`name` -> `outputName`

BruceDai replied:
Updated.

huningxin commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,19 @@
+  for (let input of resources.inputs) {
+    inputOperands.push(builder.input(input.name, {type: input.type, dimensions: input.shape}));
+  }
+  return builder.concat(inputOperands, resources.axis);
```
Consider returning the `namedOperands`.

BruceDai replied:
Updated, thanks.

BruceDai commented:
@huningxin I updated to fix your last comments, PTAL, thanks.

huningxin commented on webnn/concat.https.any.js:
```diff
@@ -0,0 +1,19 @@
+  for (let input of resources.inputs) {
+    inputOperands.push(builder.input(input.name, {type: input.type, dimensions: input.shape}));
+  }
+  return {outputOperand: builder.concat(inputOperands, resources.axis)};
```
Consider setting the output name from resources, e.g., `resources.expected.name`, instead of using a hardcoded `outputOperand` here. That would ensure the output name is consistent with `buildGraph()`.

huningxin commented on webnn/resources/utils.js:
```diff
@@ -1,61 +1,308 @@
+    const outputName = resources.expected.name ? resources.expected.name : 'outputOperand';
+    // matmul 1D with 1D produces a scalar which doesn't have its shape
+    const shape = resources.expected.shape ? resources.expected.shape : [1];
+    outputs[outputName] = new TypedArrayDict[resources.expected.type](sizeOfShape(shape));
```
The `outputName` should be consistent with operand build function, e.g., `buildConcat()`. Consider using `resources.expected.name` and adding the expected name into JSON file for each test case.

BruceDai commented:
@huningxin Updated, please take another look, thanks.

huningxin reviewed (approved):
Looks good, thanks @BruceDai !

BruceDai commented:
@dontcallmedom @Honry, I rebased with latest code, all checks have passed, PTAL, thanks.
